File: Functions/donarummo_classification.m
function minerals = donarummo_classification(peak_intensity_table)
%Mineral classification scheme from Donarummo et al. (2003)
%
%DESCRIPTION
% This function automates the mineral classification workflow published
% by Donarummo et al. (2003). Takes a table of energy dispersive spectro-
% metry (EDS) net intensity data and assigns a mineralogy to each row.
%
%SYNTAX
% minerals = DONARUMMO_CLASSIFICATION(peak_intensity_table)
%
%INPUT
% peak_intensity_table - Table containing a column for each of the
%  following elements: Na, Mg, Al, Si, K, Ca, and Fe. The name of each
%  column may be the full element name or its abbreviation. For instance,
%  "Silicon" and "Si" are valid table variable names. Both the American and
%  British spelling of "Aluminum" ("Aluminium") are also valid.
%  Capitalization is not required, but spelling is paramount. The values in
%  the table should represent the measured net intensity for each element.
%
%OUTPUT
% minerals - Categorical vector of mineral names corresponding to each row
%  in the input table. Mineral names are given as abbreviations using the
%  taxonomy proposed by Whitney and Evans (2010)*. Unknown mineral
%  classifications are given names beginning with "U-" as described by
%  Donarummo et al. (2003).
% 
%*The sixteen possible mineral classifications are:
%  ABBREVIATION     NAME
%  Ab               Albite
%  Afs              Alkali feldspar (orthoclase)
%  An               Anorthite
%  Aug              Augite
%  Bt               Biotite
%  Chl              Chlorite
%  Hbl              Hornblende
%  Htr**            Hectorite
%  Ilt              Illite
%  Ilt/Sme          Illite/Smectite 70/30 mix.
%  Kln              Kaolinite
%  Lab/Byt**        Labradorite/Bytownite
%  Mnt              Ca-Montmorillonite
%  Ms               Muscovite
%  Olig/Ans**       Oligoclase/Andesine
%  Vrm              Vermiculite
%
%**Abbreviation defined for this function.
%
%LIMITATIONS
% This function will misclassify any mineral not present in the list above.
% For instance, net intensity data for the mineral quartz will never be
% classified as quartz. To maximize the usefulness of this algorithm the
% user should also consult the results of additional classification
% methods.
%
%REFERENCES
% Donarummo, J., Ram, M., & Stoermer, E. F. (2003). Possible deposit of
%  soil dust from the 1930 s U.S. dust bowl identified in Greenland ice.
%  Geophysical Research Letters, 30(6). https://doi.org/10.1029/2002GL016641
% Whitney, D. L., & Evans, B. W. (2010). Abbreviations for names of
%  rock-forming minerals. American Mineralogist, 95(1), 185–187.
%  https://doi.org/10.2138/am.2010.3371
%
%See also
% eds_classification, kandler_classification, panta_classification, weber_classification

%Updates
% 20/Jul/2024 - Changed mineral IDs from full names to abbreviations and
% added a local function to check that the variable names in the input
% table match the required elements. Additional documentation updates.

% Function code ©2024 Austin M. Weber

%
% BEGIN FUNCTION BODY
%
% Perform initial function checks
if ~istable(peak_intensity_table)
	error('Input must be a table.')
end

% Does the table contain the correct variable names?
peak_intensity_table = parseTableVariableNames(peak_intensity_table); % Local function, defined at bottom of file

% Classify the mineral for each row of the input table
num_classifications = size(peak_intensity_table,1); 
minerals = categorical([]);
for n = 1:num_classifications
	mineral = node1(peak_intensity_table(n,:)); % Begin classification at first node of the sorting scheme tree
	minerals(n) = mineral;
end
minerals = minerals'; % Convert to column vector
% That's it! (Of course, there are a LOT of local functions behind the magic)

%
% LOCAL FUNCTIONS
%
function val = node1(T)
	ratio1 = T.Al ./ T.Si;
	if ratio1 < 0.1
		val = node2A(T);
	elseif ratio1 >= 0.7
		val = node2C(T);
	else
		val = node2B(T);
	end
end

function val = node2A(T)
	ratio2A = T.Fe ./ T.Si;
	if ratio2A >= 0.02
		val = node3A(T);
	else
		val = categorical("Htr"); % Hectorite, does not have an abbreviation in Whitney & Evans 2010
	end
end

function val = node3A(T)
	ratio3A = T.K ./ T.Al;
	if ratio3A < 0.3
		val = categorical("Aug"); % Augite
	elseif ratio3A > 0.49
		val = categorical("Hbl"); % Horblende
	else
		val = categorical("U-A");
	end
end

function val = node2C(T)
	ratio2C = (T.Mg + T.Fe) ./ T.Si;
	if ratio2C >= 0.9
		val = categorical("Chl"); % Chlorite
	elseif ratio2C < 0.3
		val = node3C(T);
	else
		val = categorical("U-E");
	end
end

function val = node3C(T)
	ratio3C = T.K ./ T.Si;
	if ratio3C >= 0.1
		val = categorical("Ms"); % Muscovite
	else
		val = node4C(T);
	end
end

function val = node4C(T)
	ratio4C = T.Ca ./ T.Si;
	if ratio4C < 0.05
		val = categorical("Kln"); % Kaolinite
	elseif ratio4C >= 0.25
		val = categorical("An"); % Anorthite
	else
		val = categorical("U-F");
	end
end

function val = node2B(T)
	ratio2B = T.K ./ (T.K + T.Na + T.Ca);
	if ratio2B < 0.35
		val = node3B1(T);
	else
		val = node3B2(T);
	end
end

function val = node3B1(T)
	ratio3B1 = (T.Mg + T.Fe) ./ T.Al;
	if ratio3B1 < 0.3
		val = node4B1a(T);
	elseif ratio3B1 >= 1.0
		val = categorical("U-C1");
	elseif (ratio3B1 > 0.5) && (ratio3B1 < 1.0)
		val = categorical("U-B1");
	else
		val = categorical("Mnt"); % Ca-Montmorillonite
	end
end

function val = node4B1a(T)
	ratio4B1a = (T.Ca + T.Na) ./ T.Al;
	if ratio4B1a >= 0.23
		val = node5B1a1(T);
	else
		val = categorical("U-B2");
	end
end

function val = node5B1a1(T)
	ratio5B1a1 = T.Ca ./ T.Na;
	if ratio5B1a1 < 0.2
		val = categorical("Ab"); % Albite
	elseif ratio5B1a1 >= 10
		val = categorical("U-B3");
	elseif (ratio5B1a1 >= 0.2) && (ratio5B1a1 < 1)
		val = categorical("Olig/Ans"); % Oligoclase/Andesine; my own abbreviations
	else
		val = categorical("Lab/Byt"); % Labradorite/Bytownite; my own abbreviations
	end
end

function val = node3B2(T)
	ratio3B2 = (T.Mg + T.Fe) ./ T.Al;
	if ratio3B2 < 0.55
		val = node4B2a(T);
	else
		val = node4B2b(T);
	end
end

function val = node4B2b(T)
	ratio4B2b = T.K ./ T.Al;
	if ratio4B2b <= 0.1
		val = categorical("U-C2");
	elseif ratio4B2b > 2
		val = categorical("Vrm"); % K-vermiculite
	elseif (ratio4B2b > 0.1) && (ratio4B2b < 1)
		val = categorical("U-D5");
	else
		val = categorical("Bt"); % Biotite
	end
end

function val = node4B2a(T)
	ratio4B2a = T.K ./ T.Al;
	if ratio4B2a >= 0.7
		val = node5B2a1(T);
	else
		val= node5B2a2(T);
	end
end

function val = node5B2a1(T)
	ratio5B2a1 = T.Al ./ T.Si;
	if ratio5B2a1 < 0.25
		val = categorical("U-D2");
	elseif (ratio5B2a1 >= 0.25) && (ratio5B2a1 <= 0.35)
		val = categorical("Afs"); % Orthoclase == alkali feldspar
	elseif (ratio5B2a1 > 0.35) && (ratio5B2a1 < 0.7)
		val = categorical("U-D1");
	else
		val = categorical("U-D5"); % This is my own classification, since Donarummo's sorting scheme does not consider classifications for any values >0.7
	end
end

function val = node5B2a2(T)
	ratio5B2a2 = T.K ./ (T.Al + T.Si);
	if ratio5B2a2 <= 0.05
		val = categorical("U-D4");
	elseif ratio5B2a2 > 0.25
		val = categorical("U-D3");
	elseif (ratio5B2a2 > 0.05) && (ratio5B2a2 <= 0.1)
		val = categorical("Ilt/Sme"); % Illite/Smectite 70/30 mixed
	else
		val = categorical("Ilt"); % Illite
	end
end

function newTable = parseTableVariableNames(originalTable)
% Checks whether the table contains the correct variables and re-names
% variables accordingly.
% Convert full-name table variables to abbreviations
element_names = lower({'Aluminum','Aluminium','Silicon',...
    'Iron','Sodium','Magnesium','Potassium','Calcium'});
varnames = originalTable.Properties.VariableNames;
varnames_lower = lower(varnames);
element_abbreviations = {'Al','Al','Si','Fe','Na','Mg','K','Ca'};
    for i = 1:numel(element_abbreviations)
	    if any(contains(varnames_lower,element_names{i}))
		    inx = contains(varnames_lower,element_names{i});
		    varnames(inx) = element_abbreviations(i);
	    end
    end
originalTable.Properties.VariableNames = varnames;
% Make sure all necessary elements are present
element_list = {'Al','Si','Fe','Na','Mg','K','Ca'};
varis = originalTable.Properties.VariableNames;
    if sum(ismember(varis,element_list)) ~= 7
	    error('Input must be a table containing columns for the elements Na, Mg, Al, Si, K, Ca, and Fe');
    end
newTable = originalTable;
end

%
% END FUNCTION BODY
%
end

File: Julia/donarummo_classification.jl
"""
Mineral classifcation scheme from Donarummo et al. (2003)

DESCRIPTION

This function automates the mineral classification workflow published by Donarummo et al. (2003). Takes a table of energy dispersive spectrometry (EDS) net intensity data and assigns a mineralogy to each row.

SYNTAX

minerals = donarummo_classification(df)

INPUT

"df" :: DataFrame containing a column for each of the following elements: Na, Mg, Al, Si, K, Ca, and Fe. The name of each column may be the full element name or its abbreviation. For instance, "Silicon" and "Si" are valid table variable names. Both the American and British spelling of "Aluminum" ("Aluminium") are also valid. Capitalization is not required by spelling is paramount. The values in the table should represent the measured net intensity for each element.

OUTPUT

"minerals" :: Categorical vector of mineral names corresponding to each row in the input DataFrame.

LIST OF POSSIBLE MINERAL CLASSIFICATIONS

 1. Albite
 2. Anorthite
 3. Augite
 4. Biotite
 5. Chlorite
 6. Hornblende
 7. Hectorite
 8. Illite
 9. Illite/Smectite 70/30 mix
10. Kaolinite
11. Labradorite/Bytownite
12. Montmorillonite
13. Muscovite
14. Oligoclase/Andesine
15. Orthoclase
16. Vermiculite

LIMITATIONS

This function will misclassify any mineral not present in the list above. For instance, net intensity data for the mineral quartz will never be classified as quartz. Some minerals may be classified as "U-" (i.e., an unknown class) if their elemental compositions do not satisify any of the indexing criteria. To maximize the usefulness of this algorithm the user should also consult the results of additional classification methods.

REFERENCE

Donarummo et al. (2003). Geophyiscal Research Letters 30(6), 1269. https://doi.org/10.1029/2002GL016641

COPYRIGHT

©2024 Austin M. Weber - function code

"""
function donarummo_classification(df)
    
        ###
        ### BEGIN FUNCTION BODY
        ###
        
        # Check that input is a DataFrame
        @assert typeof(df) == DataFrame "Input must be a DataFrame"
    
        # Convert full-name table variables to abbreviations
        element_names = ["Aluminum","Aluminium","Silicon","Iron","Sodium","Magnesium","Potassium","Calcium"] # Vector{String}
        element_names = map(lowercase,element_names)
        varnames = names(df)
        varnames_lower = map(lowercase,varnames) # Vector{String}
        element_abbreviations = ["Al","Al","Si","Fe","Na","Mg","K","Ca"]
        for (n, abbreviation) in enumerate(element_abbreviations)
            if any(x -> contains(x, element_names[n]), varnames_lower)
                idx = contains.(varnames_lower, element_names[n])
                varnames[idx] .= abbreviation
            end
        end
        rename!(df,varnames) # Replace variable names in DataFrame
    
        # Ensure that all 7 of the necessary columns exist
        if sum(in.(varnames, Ref(["Na","Mg","Al","Si","K","Ca","Fe"]))) != 7
            error("Input must be a DataFrame containing a column for Na, Mg, Al, Si, K, Ca, and Fe. Check the spellings of the column names. Only full element names and abbreviations are valid.")
        end
    
        ###
        ### CREATE LOCAL FUNCTIONS
        ###
        function mineral_classification(DF)
            num_classifications = nrow(DF)
            minerals = fill("Unknown", num_classifications) # Preallocate memory
    
            # FIRST BRANCHING PATHWAY
                # HECTORITE
                    idx = check_hectorite(DF)
                    minerals[idx] .= ["Hectorite"]
                # AUGITE
                    idx = check_augite(DF)
                    minerals[idx] .= ["Augite"]
                # U-A
                    idx = check_UA(DF)
                    minerals[idx] .= ["U-A"]
                # HORNBLENDE
                    idx = check_hornblende(DF)
                    minerals[idx] .= ["Hornblende"]
            # SECOND BRANCHING PATHWAY - B
                # ALBITE
                    idx = check_albite(DF)
                    minerals[idx] .= ["Albite"]
                # OLIG/ANDESINE
                    idx = check_olig(DF)
                    minerals[idx] .= ["Olig/Andesine"]
                # LAB/BYTOWNITE
                    idx = check_lab(DF)
                    minerals[idx] .= ["Lab/Bytownite"]
                # U-B3
                    idx = check_UB3(DF)
                    minerals[idx] .= ["U-B3"]
                # U-B2
                    idx = check_UB2(DF)
                    minerals[idx] .= ["U-B2"]
                # Ca-MONTMORILLONITE
                    idx = check_montmorillonite(DF)
                    minerals[idx] .= ["Ca-Montmorillonite"]
                # U-B1
                    idx = check_UB1(DF)
                    minerals[idx] .= ["U-B1"]
                # U-C1
                    idx = check_UC1(DF)
                    minerals[idx] .= ["U-C1"]
            # SECOND BRANCHING PATHWAY - B
                # U-D2
                    idx = check_UD1(DF)
                    minerals[idx] .= ["U-D2"]
                # ORTHOCLASE
                    idx = check_orthoclase(DF)
                    minerals[idx] .= ["Orthoclase"]
                # U-D1
                    idx = check_UD1(DF)
                    minerals[idx] .= ["U-D1"]
                # U-D0 (Note: this class is NOT defined by Donarummo et al. (2003)
                    idx = check_UD0(DF)
                    minerals[idx] .= ["U-D0"]
                # U-D4
                    idx = check_UD4(DF)
                    minerals[idx] .= ["U-D4"]
                # I/S MIXED (70/30)
                    idx = check_ISmix(DF)
                    minerals[idx] .= ["I/S Mixed (70/30)"]
                # ILLITE
                    idx = check_illite(DF)
                    minerals[idx] .= ["Illite"]
                # U-D3
                    idx = check_UD3(DF)
                    minerals[idx] .= ["U-D3"]
                # U-C2
                    idx = check_UC2(DF)
                    minerals[idx] .= ["U-C2"]
                # U-D5
                    idx = check_UD5(DF)
                    minerals[idx] .= ["U-D5"]
                # BIOTITE
                    idx = check_biotite(DF)
                    minerals[idx] .= ["Biotite"]
                # K-VERMICULITE
                    idx = check_vermiculite(DF)
                    minerals[idx] .= ["K-Vermiculite"]
            # THIRD BRANCHING PATHWAY
                # CHLORITE
                    idx = check_chlorite(DF)
                    minerals[idx] .= ["Chlorite"]
                # U-E
                    idx = check_UE(DF)
                    minerals[idx] .= ["U-E"]
                # MUSCOVITE
                    idx = check_muscovite(DF)
                    minerals[idx] .= ["Muscovite"]
                # KAOLINITE
                    idx = check_kaolinite(DF)
                    minerals[idx] .= ["Kaolinite"]
                # U-F
                    idx = check_UF(DF)
                    minerals[idx] .= ["U-F"]
                # ANORTHITE
                    idx = check_anorthite(DF)
                    minerals[idx] .= ["Anorthite"]    
            return minerals
        end ### END mineral_classification() FUNCTION
    
        # FIRST BRANCHING PATHWAY
        function check_hectorite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:Fe] ./ T[:,:Si]
            index = (criteria1 .< 0.1) .& (criteria2 .< 0.02)
            return index
        end
        function check_augite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:Fe] ./ T[:,:Si]
            criteria3 = T[:,:K] ./ T[:,:Al]
            index = (criteria1 .< 0.1) .& (criteria2 .>= 0.02) .& (criteria3 .< 0.3)
            return index
        end
        function check_UA(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:Fe] ./ T[:,:Si]
            criteria3 = T[:,:K] ./ T[:,:Al]
            index = (criteria1 .< 0.1) .& (criteria2 .>= 0.02) .& (criteria3 .>= 0.3) .& (criteria3 .<= 0.49)
            return index
        end
        function check_hornblende(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:Fe] ./ T[:,:Si]
            criteria3 = T[:,:K] ./ T[:,:Al]
            index = (criteria1 .< 0.1) .& (criteria2 .>= 0.02) .& (criteria3 .> 0.49)
            return index
        end
    
        # SECOND BRACHING PATHWAY - A
        function check_albite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = (T[:,:Ca] .+ T[:,:Na]) ./ T[:,:Al]
            criteria5 = T[:,:Ca] ./ T[:,:Na]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .< 0.35)
                .& (criteria3 .< 0.3)
                .& (criteria4 .>= 0.23)
                .& (criteria5 .< 0.2))
            return index
        end
        function check_olig(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = (T[:,:Ca] .+ T[:,:Na]) ./ T[:,:Al]
            criteria5 = T[:,:Ca] ./ T[:,:Na]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7)
                .& (criteria2 .< 0.35)
                .& (criteria3 .< 0.3)
                .& (criteria4 .>= 0.23)
                .& (criteria5 .>= 0.2) .& (criteria5 .< 1.0))
            return index
        end
        function check_lab(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = (T[:,:Ca] .+ T[:,:Na]) ./ T[:,:Al]
            criteria5 = T[:,:Ca] ./ T[:,:Na]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .< 0.35)
                .& (criteria3 .< 0.3)
                .& (criteria4 .>= 0.23)
                .& (criteria5 .>= 1.0) .& (criteria5 .< 10.0))
            return index
        end
        function check_UB3(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = (T[:,:Ca] .+ T[:,:Na]) ./ T[:,:Al]
            criteria5 = T[:,:Ca] ./ T[:,:Na]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .< 0.35)
                .& (criteria3 .< 0.3)
                .& (criteria4 .>= 0.23)
                .& (criteria5 .> 10.0))
            return index
        end
        function check_UB2(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = (T[:,:Ca] .+ T[:,:Na]) ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .< 0.35)
                .& (criteria3 .< 0.3)
                .& (criteria4 .< 0.23))
            return index
        end
        function check_montmorillonite(T)
           criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .< 0.35)
                .& (criteria3 .>= 0.3) .& (criteria3 .<= 0.5))
            return index
        end
        function check_UB1(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .< 0.35)
                .& (criteria3 .> 0.5) .& (criteria3 .< 1.0))
            return index
        end
        function check_UC1(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .< 0.35)
                .& (criteria3 .>= 1.0))
            return index
        end
        
        # SECOND BRANCHING PATHWAY - B
        function check_UD2(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:Al] ./ T[:,:Si]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .>= 0.7)
                .& (criteria5 .< 0.25))
            return index
        end
        function check_orthoclase(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:Al] ./ T[:,:Si]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .>= 0.7)
                .& (criteria5 .>= 0.25) .& (criteria5 .< 0.35))
            return index
        end
        function check_UD1(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:Al] ./ T[:,:Si]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .>= 0.7)
                .& (criteria5 .> 0.35) .& (criteria5 .< 0.7))
            return index
        end
        function check_UD0(T)
            # Note: This class is NOT defined in Donarummo et al. (2003)!
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:Al] ./ T[:,:Si]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .>= 0.7)
                .& (criteria5 .>= 0.7))
            return index
        end
        function check_UD4(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:K] ./ (T[:,:Al] .+ T[:,:Si])
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .< 0.7)
                .& (criteria5 .<= 0.05))
            return index
        end
        function check_ISmix(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:K] ./ (T[:,:Al] .+ T[:,:Si])
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .< 0.7)
                .& (criteria5 .> 0.05) .& (criteria5 .<= 0.1))
            return index
        end
        function check_illite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:K] ./ (T[:,:Al] .+ T[:,:Si])
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .< 0.7)
                .& (criteria5 .> 0.1) .& (criteria5 .<= 0.25))
            return index
        end
        function check_UD3(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            criteria5 = T[:,:K] ./ (T[:,:Al] .+ T[:,:Si])
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .< 0.7)
                .& (criteria5 .> 0.25))
            return index
        end
        function check_UC2(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .<= 0.1))
            return index
        end
        function check_UD5(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .> 0.1) .& (criteria4 .< 1.0))
            return index
        end
        function check_biotite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .>= 1.0) .& (criteria4 .<= 2.0))
            return index
        end
        function check_vermiculite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = T[:,:K] ./ (T[:,:K] .+ T[:,:Na] .+ T[:,:Ca])
            criteria3 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Al]
            criteria4 = T[:,:K] ./ T[:,:Al]
            index = ((criteria1 .>= 0.1) .& (criteria1 .< 0.7) 
                .& (criteria2 .>= 0.35)
                .& (criteria3 .< 0.55)
                .& (criteria4 .> 2.0))
            return index
        end
    
        # THIRD BRANCHING PATHWAY
        function check_chlorite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Si]
            index = (criteria1 .>= 0.7) .& (criteria2 .>= 0.9)
            return index
        end
        function check_UE(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Si]
            index = (criteria1 .>= 0.7) .& (criteria2 .>= 0.3) .& (criteria2 .< 0.9) 
            return index
        end
        function check_muscovite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Si]
            criteria3 = T[:,:K] ./ T[:,:Si]
            index = (criteria1 .>= 0.7) .& (criteria2 .< 0.3) .& (criteria3 .>= 0.1) 
            return index
        end
        function check_kaolinite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Si]
            criteria3 = T[:,:K] ./ T[:,:Si]
            criteria4 = T[:,:Ca] ./ T[:,:Si]
            index = ((criteria1 .>= 0.7) 
                .& (criteria2 .< 0.3) 
                .& (criteria3 .< 0.1) 
                .& (criteria4 .< 0.05))
            return index
        end
        function check_UF(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Si]
            criteria3 = T[:,:K] ./ T[:,:Si]
            criteria4 = T[:,:Ca] ./ T[:,:Si]
            index = ((criteria1 .>= 0.7) 
                .& (criteria2 .< 0.3) 
                .& (criteria3 .< 0.1) 
                .& (criteria4 .>= 0.05) .& (criteria4 .< 0.25))
            return index
        end
        function check_anorthite(T)
            criteria1 = T[:,:Al] ./ T[:,:Si]
            criteria2 = (T[:,:Mg] .+ T[:,:Fe]) ./ T[:,:Si]
            criteria3 = T[:,:K] ./ T[:,:Si]
            criteria4 = T[:,:Ca] ./ T[:,:Si]
            index = ((criteria1 .>= 0.7) 
                .& (criteria2 .< 0.3) 
                .& (criteria3 .< 0.1) 
                .& (criteria4 .>= 0.25))
            return index
        end
        
        # Final step: Classify the mineralogy for each row of the input table
        minerals = DataFrame(Minerals = mineral_classification(df))
        
        return minerals
end